### ISSUE TITLE
uulib hangs with GHC 9.8

### ISSUE THREAD

**Issue by jbransen** [UU-ComputerScience/uulib#11]
See https://github.com/UU-ComputerScience/uuagc/issues/15 for more info, the following example hangs with GHC 9.8 while it works with earlier versions:

```
module Main where

import UU.Parsing
import UU.Scanner

type TokenParser a = Parser Token a

parseTokens :: TokenParser a -> [Token] -> Either [String] a
parseTokens p tks
  = if null msgs
    then final `seq` Right v
    else Left (map show msgs)
  where
    steps = parse p tks
    msgs  = getMsgs steps
    (Pair v final) = evalSteps steps

main :: IO ()
main
  = let res = parseTokens pVarid [reserved "x" noPos]
    in case res of
         Left errs -> mapM_ putStrLn errs
         Right tree -> putStrLn $ show tree
```

**Comment by jbransen** [comment]
Compiling `uulib` with profiling on removes the loop. Great...

**Comment by jbransen** [comment]
I compiled with `--ghc-options="-g" --no-executable-stripping` and used [perf](https://www.schoolofhaskell.com/user/bitonic/perf-for-low-level-profiling) for profiling, this gives the following report:
```
# Overhead  Command          Shared Object                       Symbol                                                                                                                                                                                                                                                                                                                           
    44.35%  uulib            uulib                               [.] evacuate
    14.17%  uulib            uulib                               [.] scavenge_block
     2.79%  uulib            uulib                               [.] scavenge_stack
     1.35%  uulib            uulib                               [.] evacuate_BLACKHOLE
     1.29%  uulib            uulib                               [.] raiseAsync
     1.22%  uulib            uulib                               [.] updateThunk
     1.10%  uulib            [kernel.kallsyms]                   [k] sync_regs
     1.05%  uulib            [kernel.kallsyms]                   [k] native_irq_return_iret
     0.96%  uulib            [kernel.kallsyms]                   [k] __rcu_read_unlock
     0.81%  uulib            [kernel.kallsyms]                   [k] __count_memcg_events
     0.80%  uulib            uulib                               [.] scavenge_mutable_list
     0.79%  uulib            [kernel.kallsyms]                   [k] get_mem_cgroup_from_mm
     0.68%  uulib            uulib                               [.] uulibzm0zi9zi24zm47fxvoHBG7EGQ6kMfQl4gf_UUziParsingziInterface_zdfApplicativeAnaParser_info
     0.63%  swapper          [kernel.kallsyms]                   [k] intel_idle
     0.56%  uulib            uulib                               [.] threadPaused
     0.55%  uulib            uulib                               [.] GarbageCollect
     0.51%  uulib            [kernel.kallsyms]                   [k] try_charge_memcg
     0.50%  uulib            uulib                               [.] todo_block_full
     0.50%  uulib            [kernel.kallsyms]                   [k] __mod_memcg_lruvec_state
     0.48%  uulib            [kernel.kallsyms]                   [k] lru_gen_add_folio
     0.48%  irq/9-acpi       [kernel.kallsyms]                   [k] acpi_os_read_port
     0.43%  uulib            uulib                               [.] scavenge_until_all_done
     0.41%  uulib            [kernel.kallsyms]                   [k] lru_gen_del_folio.constprop.0
     0.41%  uulib            [kernel.kallsyms]                   [k] cgroup_rstat_updated
     0.37%  swapper          [kernel.kallsyms]                   [k] _nv039899rm
     0.35%  uulib            uulib                               [.] freeGroup
     0.34%  uulib            [kernel.kallsyms]                   [k] clear_page_erms
     0.32%  uulib            [kernel.kallsyms]                   [k] __handle_mm_fault
     0.30%  uulib            [kernel.kallsyms]                   [k] lru_lazyfree_fn.part.0
     0.28%  uulib            [kernel.kallsyms]                   [k] release_pages
     0.26%  swapper          [kernel.kallsyms]                   [k] xhci_get_isoc_frame_id
     0.25%  uulib            [kernel.kallsyms]                   [k] xhci_get_isoc_frame_id
     0.25%  uulib            uulib                               [.] stg_upd_frame_info+0xffffffffffc0001e
     0.24%  uulib            [kernel.kallsyms]                   [k] mas_walk
     0.24%  uulib            [kernel.kallsyms]                   [k] __mod_node_page_state
     0.24%  uulib            uulib                               [.] alloc_todo_block
     0.24%  uulib            uulib                               [.] scavenge_one
```

in other words, the process is busy garbage collecting. `uulibzm0zi9zi24zm47fxvoHBG7EGQ6kMfQl4gf_UUziParsingziInterface_zdfApplicativeAnaParser_info` is likely the place where allocations happen.

**Comment by noughtmare** [comment]
I'm very surprised that the class constraint causes an infinite loop. Do you know why that was happening? Is it using a different instance now?

**Comment by noughtmare** [comment]
Oh, it was using UndecidableInstances...

**Comment by jbransen** [comment]
I did not investigate too much, but since https://wiki.haskell.org/Functor-Applicative-Monad_Proposal we have `Functor f => Applicative f`, and then in this instance we have `Applicative f => Functor f`, so that's a self-dependency on the class constraint level. And with `UndecidableInstances` we apparently had bad luck this time with the way these constraints are solved. However, the explicit constraint has not been needed since many GHC versions, so I decided to just remove it.
